text/0082-dynamic-size-region.md

```diff
@@ -0,0 +1,129 @@
+to PD. For hotspot regions, buckets are split by scan. For cold regions, buckets are split by approximate
+range size. Every bucket should have the size about 128MiB. Their ranges and stats are all reported to PD. PD
+can detect hot buckets from all regions and decide whether need to split new hotspots and schedules them to
+balance.
+
+![dynamic size buckets](../media/dynamic-size-buckets.png)
+
+### Read and Write
+
+Since regions’ sizes become dynamic, region count in a single instance is not a problem any more.
+Hibernate regions can be deprecated, and lease renew by heartbeats can be enabled by default for
+all regions. This can bring significant improvement for read tail latency.
+
+After the region becomes larger, the chance to hit 1PC becomes higher. And the RPC count in a
+transaction can also be reduced. However, the hotspot chance is also increased. For writes, we rely
+on PD to split the hotspot in time and increase concurrency. In addition, we can also use unorder
+apply to make single region apply logs faster. Since unorder apply is a standalone feature, I’m not
+going into details here.
+
+For read hotspots, split should be triggered by PD, which can utilize the global statics from all
+regions and nodes. For normal read requests, TiKV will need to split its range into smaller buckets
+according to statics to increase concurrency. When TiDB wants to do a scan, it sends the RPC once,
+and TiKV will split the requests into smaller concurrent jobs.
+
+In the past design, follower read has also been discussed to offload works from leader. But TiDB
+can’t predict the progress of a follower, so latency can also become unpredictable. It’s more
+controllable to let PD split hotspots and schedule to get balance.
+
+Now that a region can be in 10GiB, a full scan can make the response exceed the limit of gRPC,
+which is 4GiB. So instead of unary, we need to use server streaming RPC to return the response as
+stream. The stream can set a limit on memory or count to avoid too many pending responses
+exhausting memory.
+
+After a region becomes larger, it needs to be warmed up before serving requests. There are two
+common cases:
+
+- In the very beginning, writes will only happen in a single region, the problem can be solved by
+    presplit + scatter.
+- Because followers don’t serve read, so their data is usually not in cache. After transferring
+    the leader, serving a range scan can trigger at most 10GiB IO. To warm it up, we can extend
+    the existing pre transfer leader mechanism. Leader will send its bucket statics to follower
+    in pre transfer leader, follower can do a pre scan for hot buckets before accepting
+    transferring leader.
+
+### Flow report
+
+Buckets statistics are reported in a standalone stream. They will be reported more frequently
+than original region heartbeats, which is 60s. The new stream is expected to be reported every
+10 seconds when have changes.
+
+### Replication
```

Comment:
Seems with TiKV's customed SstPartitioner, a large region can cause a large SST file? It can increase write amplification, so I think we also need to update the customed SstPartitioner.

Reply:
Why SstPartitioner can cause large SST file? I think the SST file size is controlled by `target_file_size_base` configuration.

Reply:
Ah I forget this.